### ISSUE TITLE
Merge layer implemented with inplace feature or not?

### ISSUE THREAD

**Issue by iammarvelous** [fchollet/keras#4991]
From this [line](https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1386), I am not sure whether it allocates a new memory for the result. If so, one can use things like [assign_add](https://www.tensorflow.org/api_docs/python/state_ops/variables#Variable.assign_add) to avoid unnecessary memory allocation. I am asking about this because in my model there are many merge layer and I found out that it consumes much more memory in keras with tensorflow backend (twice as much that in Torch). Besides, GPU usage is at 50%-60% during training. I am doubting the memory allocation could be the bottleneck. Correct me if I misunderstood.

**Comment by bstriner** [comment]
The highlighted line checks if mode is a callable and if that callable has an argument named mask, then pass mask to the callable. Did the line numbers shift or something?

Merge layer just tells the backend to perform a calculation with multiple inputs. How the backend optimizes those calculations is up to the backend, but in general, it isn't doing any unnecessary memory allocations.

Keras code is mostly building a function graph that is then optimized and run. The code in "call" is run once and builds a graph that is run later. It isn't literally performing the computation. Please post in theano or tensorflow use groups if you have more questions about how the backend works.

Cheers,
Ben

**Comment by iammarvelous** [comment]
Oh, the line got shifted somehow. I was pointing this [line](https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L1386). By the way, I am quite confused about keras tensor operators since I did not find any document/code that overload operators like `=` or `+=`. By saying unnecessary memory allocation, I wonder whether `Merge([a,b], mode='sum')` create a new tensor to hold the sum of `a` and `b` or it is actually like something `a = a + b` which put results in one of the original tensor.

**Comment by bstriner** [comment]
The answer is neither. If a and b are tensors, `c=a+b` creates a function graph that will add two variables and return a third variable. `d=2*c` will multiply the result by two.

After you make the graph, you say `f=K.function([a,b],d)`. This will create the function you actually use. It will have two inputs and output 2*(input1+input2).

This is kind of the beauty of theano/tensorflow. You create the mathematical graph and tell it what inputs and outputs you want, and it will find the most efficient way to do so.

If you're having memory issues, it's something else in what you're trying to do, but for the most part the backend is incredibly efficient.

Regarding how it all works, search theano or tensorflow for `__add__`, `__mul__`, etc.

**Comment by iammarvelous** [comment]
I get it, thanks. There seems to be an issue for keras (maybe this discussion is off the original topic). I guess in this case keras just assume those basic operators (`__add__`) have similar broadcasting behavior in both theano and tensorflow. But I have come across a case earlier that the behaviors of those two backends on basic operators are not quite same. Here is an example, when `a` is a matrix and `b` is a vector. Theano backend would broadcast but tensorflow would not do that. Right now, I use the following way to work around.

```
def k_div(x, y):
    try:
        return x/y
    except:
        import tensorflow as tf
        return tf.div(x, y)
```

**Comment by iammarvelous** [comment]
By the way, memory issue is not my biggest concern. I just feel that 50% - 60% GPU usage is weird and inefficient, and I am trying to figure out the reason for this. Did you ever meet any similar problem?

**Comment by bobye** [comment]
@bstriner I don't think tensor has the capability to automatically resolve such memory footprint for tensor operations. For example, it has `.assign_add()` operator that allows user to specify the how the calculation is done in memory.  https://www.tensorflow.org/api_docs/python/state_ops/variables#Variable.assign_add

**Comment by iammarvelous** [comment]
@bstriner Well, I don't think tensorflow handles memory automatically. It is up to how you define the graph, otherwise there should be no difference between `tf.__add__` and `tf.assign_add`.

**Comment by bstriner** [comment]
In terms of memory, theano and tensorflow are probably streaming all of the intermediate variables or otherwise conserving memory.

Tensorflow doesn't have some of the graph optimization that theano does, which is kind of a shame. I'd be interested in how your usage changes on theano.

There is constant discussion around implementing graph optimizers for tensorflow. Wouldn't be that hard to get started. Imagine a rule that replaces assign(x,x+d) with assign_add(x,d).

If you're crushing your GPU, the most likely cause is that you're asking it to do something complicated or giving it batches that are too large. Depending on what you're trying to do, that is what you expect to happen. What is your total batch input and target size?

**Comment by iammarvelous** [comment]
@bstriner I attempted to use theano backend before. However, it took too long (more than 1 hour) to compile the model and I was not patient enough to wait for it.

I have set batch size to 4 and it would consume almost all 16 GPU memory on P100. It would not crush at size 4 and one epoch takes about 1 hour. I have implemented the exact same model in Torch. I can set batch size to 8 and one epoch takes about 15-20 min. The running time is the biggest concern for me.

**Comment by bstriner** [comment]
Theano can take a crazy long time to compile everything, especially the first time it is using certain operations. However, part of that is because it is trying to perform graph optimizations by merging redundant calculations and whatever else it does. So I would be interested to see the actual running time with theano.

This is what tensorflow needs to implement badly. There are some plans but nothing working yet.
http://deeplearning.net/software/theano/extending/optimization.html

Is your data insanely wide or something? I don't get how a batch size of 4 consumes everything. Are you making like a really deep LSTM or something? Sorry, but not quire sure with what you're working on yet.

**Comment by iammarvelous** [comment]
@bstriner I have just tried theano backend. It seems that the per epoch training time goes to much much longer (almost looks like cpu training, but I think I did set it up correctly).

My data is just 256x256x3 image and no recurrent stuff is involved. Basically, I reimplement [this](https://github.com/anewell/pose-hg-train) in keras.
